### PR TITLE
feat: add use case for refilling key packages

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- Add a use case for refilling the key packages if are 50% below the desired count.
- Add CLI command for refilling key packages

https://wearezeta.atlassian.net/browse/CL-10

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Change the amount key packages we initially upload to something like 10.

run:
`> cli login refill-key-packages`

And you should observe that we initially upload 10 key packages followed by 90 when we check if we need to refill them.

### Notes

As discussed the refilling could be triggered in a sync step. This will be done in separate PR.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+class RefillKeyPackagesUseCaseImpl(
+    private val keyPackageRepository: KeyPackageRepository,
+    private val clientRepository: ClientRepository
+) : RefillKeyPackagesUseCase {
+    override suspend operator fun invoke(): RefillKeyPackagesResult =
+
+        clientRepository.currentClientId().flatMap { selfClientId ->
+            keyPackageRepository.getAvailableKeyPackageCount(selfClientId)
+                .flatMap { count ->
+                    if (count.count < (KEY_PACKAGE_LIMIT * 0.5)) {
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt:34]: can we have this calculation extracted as a function and also that 0.5 number as a variable that clear the intention.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    @BeforeTest
+    fun setup() {
+        refillKeyPackageUseCase = RefillKeyPackagesUseCaseImpl(
+            keyPackageRepository, clientRepository
+        )
+    }
+
+    @Test
+    fun givenKeyPackageCountIs50PercentBelowLimit_ThenRequestToRefillKeyPackageIsPerformed() = runTest {
```
[the remark this answers is not in the document]
Comment (typfel) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt:41]: It's tested in `givenKeyPackageCount50PercentAboveLimit_ThenNoRequestToRefillKeyPackagesIsPerformed `

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        val keyPackageCount = (KEY_PACKAGE_LIMIT * 0.5 - 1).toInt()
+
+        given(clientRepository).suspendFunction(clientRepository::currentClientId).whenInvoked().then { Either.Right(TestClient.CLIENT_ID) }
```
Comment (mchenani) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt:44]: Could you please check other tests and the arrangement  practice in cleaning the test classes? Please take a look at AddMemberToConversationUseCaseTest for example.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt (hunk 2 above)
Comment (mchenani) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt:41]: Just saw it, deleted my comment. Thanks

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt (hunk 3 above)
Comment (typfel) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCaseTest.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt (hunk 1 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+class RefillKeyPackagesUseCaseImpl(
+    private val keyPackageRepository: KeyPackageRepository,
+    private val clientRepository: ClientRepository
+) : RefillKeyPackagesUseCase {
+    override suspend operator fun invoke(): RefillKeyPackagesResult =
+
+        clientRepository.currentClientId().flatMap { selfClientId ->
+            keyPackageRepository.getAvailableKeyPackageCount(selfClientId)
+                .flatMap { count ->
```
[the remark this answers is not in the document]
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/RefillKeyPackageUseCase.kt]: I removed the name, are you OK with that?

**Pull request merged**